### PR TITLE
[OMID-78] Identify transaction snapshot at the server

### PR DESCRIPTION
**Pull request by ohadshacham**
To identify a transaction's snapshot in scans and gets, Omid reads and filters cells at the client side. This filtering requires checking the shadow cells and accessing the commit table if needed.
Doing the filtering at the server side should be more efficient since less data is being transferred to the client and also, in some cases, RPCs are being saved.
This feature requires implementing a coprocessor that will do the filtering at the server side and returns only the relevant snapshot to the client side.

This feature is also needed to support the integration of Omid with Phoenix. This is required since Phoenix has coprocessors that change and condence the data returned to the client and therefore, filtering at the client side in this case is impossible.

### CODE REVIEW

File: hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        this.maxVersions = maxVersions;
+        this.familyDeletionCache = new HashMap<String, List<Cell>>();
+    }
+
+    @Override
+    public boolean next(List<Cell> results) throws IOException {
+       return next(results, Integer.MAX_VALUE);
+    }
+
+    public boolean next(List<Cell> result, int limit) throws IOException {
```
Comment (ebortnik) [inline hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java]: raw --> row

File: hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        return scanner.getMaxResultSize();
+    }
+
+    @Override
+    public long getMvccReadPoint() {
+        return scanner.getMvccReadPoint();
+    }
+
+    @Override
+    public boolean nextRaw(List<Cell> result) throws IOException {
```
Comment (ebortnik) [inline hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java]: raw --> row

File: hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/RegionAccessWrapper.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import java.io.IOException;
+import java.util.List;
+
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.HTableInterface;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.omid.transaction.TableAccessWrapper;
+
+public class RegionAccessWrapper implements TableAccessWrapper {
```
Comment (ebortnik) [inline hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/RegionAccessWrapper.java]: High level comment please

File: hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.HTableInterface;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+
+public class HTableAccessWrapper implements TableAccessWrapper {
```
Comment (ebortnik) [inline hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java]: What are other implementations of this interface?

File: hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    private final HTableInterface healerTable;
+    private HTableInterface table;
```
Comment (ebortnik) [inline hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java]: Why not final?

File: hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.HTableInterface;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+
+public class HTableAccessWrapper implements TableAccessWrapper {
+
+    private final HTableInterface healerTable;
```
Comment (ebortnik) [inline hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java]: Consider other names, e.g., healerTable --> wTable, table --> rTable

File: hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java (hunk 1 above)
Comment (ohadshacham) [reply hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java]: The function is defined in RegionScanner interface, we are just override it.

File: hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java (hunk 2 above)
Comment (ohadshacham) [reply hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java]: The function is defined in RegionScanner interface, we are just override it.

File: hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java (hunk 4 above)
Comment (ohadshacham) [reply hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java]: RegionAccessWrapper

File: hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/RegionAccessWrapper.java (hunk 3 above)
Comment (ohadshacham) [reply hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/RegionAccessWrapper.java]: done

File: hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java (hunk 6 above)
Comment (ohadshacham) [reply hbase-client/src/main/java/org/apache/omid/transaction/HTableAccessWrapper.java]: Done

File: hbase-client/src/main/java/org/apache/omid/transaction/TTable.java [hunk 7]
```diff
@@ -76,6 +71,10 @@
 
     private HTableInterface table;
 
+    private SnapshotFilter snapshotFilter;
+
+    private boolean serverSideFilter;
```
Comment (yonigottesman) [inline hbase-client/src/main/java/org/apache/omid/transaction/TTable.java]: No need to be a class field, you use it once in a function so make it local there.

File: hbase-client/src/main/java/org/apache/omid/transaction/TTable.java [hunk 8]
```diff
@@ -92,16 +95,40 @@ public TTable(Configuration conf, String tableName) throws IOException {
...
         table = hTable;
         healerTable = new HTable(table.getConfiguration(), table.getTableName());
+        serverSideFilter = table.getConfiguration().getBoolean("omid.server.side.filter", false);
+        snapshotFilter = (serverSideFilter) ?  new AttributeSetSnapshotFilter(hTable) : new SnapshotFilterImpl(new HTableAccessWrapper(hTable, healerTable));
+    }
+
+    public TTable(HTableInterface hTable, CommitTable.Client commitTableClient) throws IOException {
+        table = hTable;
+        healerTable = new HTable(table.getConfiguration(), table.getTableName());
+        serverSideFilter = table.getConfiguration().getBoolean("omid.server.side.filter", false);
```
Comment (yonigottesman) [inline hbase-client/src/main/java/org/apache/omid/transaction/TTable.java]: serverSideFilter should be local

File: hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.HRegionInfo;
+import org.apache.hadoop.hbase.regionserver.RegionScanner;
+import org.apache.hadoop.hbase.regionserver.ScannerContext;
+import org.apache.omid.transaction.HBaseTransaction;
+import org.apache.omid.transaction.SnapshotFilterImpl;
+
+public class OmidRegionScanner implements RegionScanner {
+
+    RegionScanner scanner;
```
Comment (yonigottesman) [inline hbase-coprocessor/src/main/java/org/apache/hadoop/hbase/regionserver/OmidRegionScanner.java]: Why are these package-private and not private?

File: hbase-client/src/main/java/org/apache/omid/transaction/HBaseTransactionManager.java [hunk 10]
```diff
@@ -292,7 +268,7 @@ public ConflictDetectionLevel getConflictDetectionLevel() {
             get.addColumn(family, shadowCellQualifier);
             get.setMaxVersions(1);
             get.setTimeStamp(startTimestamp);
-            Result result = hBaseCellId.getTable().get(get);
+            Result result = (tableAccessWrapper != null) ? tableAccessWrapper.get(get) : hBaseCellId.getTable().get(get);
```
Comment (yonigottesman) [inline hbase-client/src/main/java/org/apache/omid/transaction/HBaseTransactionManager.java]: why should this be null? Maybe its better to pass "hBaseCellId.getTable().get(get);" to the constructor code

File: hbase-client/src/main/java/org/apache/omid/transaction/SnapshotFilter.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+public interface SnapshotFilter {
+    
+    public Result get(TTable ttable, Get get, HBaseTransaction transaction) throws IOException;
+
+    public ResultScanner getScanner(TTable ttable, Scan scan, HBaseTransaction transaction) throws IOException;
+
+    public List<Cell> filterCellsForSnapshot(List<Cell> rawCells, HBaseTransaction transaction,
+            int versionsToRequest, Map<String, List<Cell>> familyDeletionCache) throws IOException;
+
+    public boolean isCommitted(HBaseCellId hBaseCellId, long epoch) throws TransactionException;
```
Comment (yonigottesman) [inline hbase-client/src/main/java/org/apache/omid/transaction/SnapshotFilter.java]: Why do all other function throw ioexceptions and this one something else?

**Pull request closed without merging**